File: validation/typescript-server/test/server.test.mjs
import assert from "node:assert/strict";
import test from "node:test";
import {
  InternalCommandSchema,
  ServerRequestContextSchema,
  TrustedActorSchema,
} from "../dist/server.js";

const validActor = {
  userId: "user-1",
  tenantId: "tenant-1",
  roles: ["reader"],
};

const validContext = {
  requestId: "req-1",
  traceId: "trace-1",
  actor: validActor,
  sourceIp: "127.0.0.1",
};

test("accepts a bounded trusted actor and request context", () => {
  assert.deepEqual(TrustedActorSchema.parse(validActor), validActor);
  assert.deepEqual(ServerRequestContextSchema.parse(validContext), validContext);
});

test("preserves exact trusted identity strings", () => {
  const actor = { userId: " user-1 ", roles: [] };
  assert.deepEqual(TrustedActorSchema.parse(actor), actor);
});

for (const [name, actor] of [
  ["empty user id", { userId: "", roles: [] }],
  ["oversized user id", { userId: "u".repeat(129), roles: [] }],
  ["empty role", { userId: "user-1", roles: [""] }],
  ["oversized role", { userId: "user-1", roles: ["r".repeat(129)] }],
  ["too many roles", { userId: "user-1", roles: Array.from({ length: 65 }, () => "reader") }],
  ["unknown actor field", { userId: "user-1", roles: [], token: "must-not-leak" }],
]) {
  test(`rejects trusted actor with ${name}`, () => {
    assert.equal(TrustedActorSchema.safeParse(actor).success, false);
  });
}

for (const [name, context] of [
  ["invalid source IP", { ...validContext, sourceIp: "not-an-ip" }],
  ["unknown public identity", { ...validContext, userId: "client-supplied" }],
  ["invalid nested request metadata", { ...validContext, requestId: "" }],
]) {
  test(`rejects server request context with ${name}`, () => {
    assert.equal(ServerRequestContextSchema.safeParse(context).success, false);
  });
}

test("accepts a complete internal command", () => {
  const command = {
    operationId: "alerts.create",
    idempotencyKey: "idem-1",
    context: validContext,
    payload: { message: "hello" },
  };
  assert.deepEqual(InternalCommandSchema.parse(command), command);
});

for (const [name, command] of [
  ["missing operation id", { context: validContext, payload: {} }],
  ["empty operation id", { operationId: "", context: validContext, payload: {} }],
  ["oversized operation id", { operationId: "o".repeat(257), context: validContext, payload: {} }],
  ["oversized idempotency key", { operationId: "alerts.create", idempotencyKey: "i".repeat(129), context: validContext, payload: {} }],
  ["unknown command field", { operationId: "alerts.create", context: validContext, payload: {}, credential: "secret" }],
]) {
  test(`rejects internal command with ${name}`, () => {
    assert.equal(InternalCommandSchema.safeParse(command).success, false);
  });
}

File: validation/golang-server/server_test.go
package servervalidation

import (
	"strings"
	"testing"

	public "github.com/fanwaave/fanwaave-lib-core/validation/golang"
)

func validContext() ServerRequestContext {
	return ServerRequestContext{
		Public: public.RequestMeta{
			RequestID: "req-1",
			TraceID:   "trace-1",
		},
		Actor: TrustedActor{
			UserID: "user-1",
			Roles:  []string{"reader"},
		},
		SourceIP: "127.0.0.1",
	}
}

func TestAcceptsBoundedServerValues(t *testing.T) {
	if err := Validate(validContext()); err != nil {
		t.Fatalf("expected valid context: %v", err)
	}
	command := InternalCommand{
		OperationID:    "alerts.create",
		IdempotencyKey: "idem-1",
		Context:        validContext(),
		Payload:        map[string]any{"message": "hello"},
	}
	if err := Validate(command); err != nil {
		t.Fatalf("expected valid command: %v", err)
	}
}

func TestRejectsInvalidTrustedActor(t *testing.T) {
	cases := []TrustedActor{
		{UserID: "", Roles: nil},
		{UserID: strings.Repeat("u", 129), Roles: nil},
		{UserID: "user-1", TenantID: strings.Repeat("t", 129), Roles: nil},
		{UserID: "user-1", Roles: []string{""}},
		{UserID: "user-1", Roles: []string{strings.Repeat("r", 129)}},
		{UserID: "user-1", Roles: make([]string, 65)},
	}
	for _, actor := range cases {
		if err := Validate(actor); err == nil {
			t.Fatalf("expected invalid actor: %#v", actor)
		}
	}
}

func TestRejectsInvalidServerContext(t *testing.T) {
	value := validContext()
	value.SourceIP = "not-an-ip"
	if err := Validate(value); err == nil {
		t.Fatal("expected invalid source IP")
	}

	value = validContext()
	value.Public.RequestID = ""
	if err := Validate(value); err == nil {
		t.Fatal("expected invalid nested public metadata")
	}
}

func TestRejectsInvalidInternalCommand(t *testing.T) {
	cases := []InternalCommand{
		{OperationID: "", Context: validContext()},
		{OperationID: strings.Repeat("o", 257), Context: validContext()},
		{OperationID: "alerts.create", IdempotencyKey: strings.Repeat("i", 129), Context: validContext()},
	}
	for _, command := range cases {
		if err := Validate(command); err == nil {
			t.Fatalf("expected invalid command: %#v", command)
		}
	}
}
